gcs/src/gcs_group.cpp

```diff
@@ -1165,7 +1185,10 @@ group_find_ist_donor (const gcs_group_t* const group,
              (long long)ist_seqno, (long long)lowest_cached_seqno,
              (long long)conf_seqno, (long long)safe_ist_seqno);
 
-    if (ist_seqno < safe_ist_seqno) {
+    // We should ignore this heuristic if the request contains only a IST part,
+    // since otherwise we cannot fulfill the request:
+
+    if (ist_seqno < safe_ist_seqno && ! ist_only) {
```

Comment:
In the first version of this patch, we have already experimented with the alternative algorithm of calculation of the safety_gap.

However, really it reduces probability of the successful completion of the IST in all other scenarios that occur continuously in real life, while problems with previous heuristic appeared only once in a long while stress testing. If we put this new heuristics, we will get this problem more often, not less often.

Therefore, I had canceled code changes that lead to more frequent hit of the scenario, when the IST is not possible and SST (if present) will be used instead (or when we need to change the donor during IST - which is not implemented yet in the Galera).